File: dataStructures/stackQeueDeque/javascript/stack.js
// Stack implemented with JS Array, using a constructor function

function Stack() {
  this.container = []
}

Stack.prototype.push = function(item) {
  this.container.push(item)
}

Stack.prototype.pop = function(idx = null) {
  if (idx) {
    const itemToReturn = this.container[idx]
    this.container = this.container.filter((item, _idx) => idx !== _idx)
    return itemToReturn
  }
  return this.container.pop()
}

Stack.prototype.size = function() {
  return this.container.length
}

Stack.prototype.isEmpty = function() {
  return !this.container.length
}

Stack.prototype.serialize = function() {
  return this.container.join("-")
}

const s = new Stack()
s.push("hello")
s.push(4)
console.log(s.size()) // 2
console.log(s.isEmpty()) // false
console.log(s.serialize()) // "hello-4"
console.log(s.pop()) // 4
console.log(s.pop()) // "hello"
console.log(s.isEmpty()) // true

File: 1-StackQueueDeque/python/stack.py
class Stack(object):
    def __init__(self):
        self.container = []

    def __len__(self):
        """
        dunder method implementation of size()
        """
        return len(self.container)

    def __str__(self):
        """
        Return stack as string, delimited by "-"
        """
        return "-".join([str(item) for item in self.container])

    def push(self, item):
        self.container.append(item)

    def pop(self, idx = None):
        if idx is None:
            return self.container.pop()
        else:
            item_to_return = self.container[idx]
            self.container = [item for i, item in enumerate(self.container) if i != idx]
            return item_to_return

    def peek(self):
        return self.container[-1]

    def is_empty(self):
        return not len(self.container)

if __name__ == "__main__":
    s = Stack()
    print(s.is_empty()) # True
    s.push(4)
    s.push('dog')
    print(s.peek()) # 'dog'
    print(s.pop()) # 'dog'
    s.push('cat')
    s.push(100)
    print(s.pop(1)) # 'cat'
    print(len(s)) # 2
    print(s) # 4-100
